### PR TITLE
Fix Posix Serial Driver Not Unblocking Read on Close

### PR DESCRIPTION
**Pull request by ryanmelt**

### CODE REVIEW

File: openc3/lib/openc3/io/posix_serial_driver.rb [hunk 1]
```diff
@@ -132,9 +137,19 @@ def read
       begin
         data = @handle.read_nonblock(65535)
       rescue Errno::EAGAIN, Errno::EWOULDBLOCK
-        result = IO.fast_select([@handle], nil, nil, @read_timeout)
-        if result
-          retry
+        begin
+          read_ready, _ = IO.fast_select(@readers, nil, nil, @read_timeout)
+        rescue IOError
+          @pipe_reader.close unless @pipe_reader.closed?
```
Comment (jmthomas) [inline openc3/lib/openc3/io/posix_serial_driver.rb:143]: I don't see how closing the pipe_reader is going to help with the serial port handle

File: openc3/lib/openc3/io/posix_serial_driver.rb [hunk 2]
```diff
@@ -132,9 +137,19 @@ def read
...
+          return ""
+        end
+        if read_ready
+          if read_ready.include?(@pipe_reader)
+            @pipe_reader.close unless @pipe_reader.closed?
+            return ""
+          else
+            retry
```
Comment (jmthomas) [inline openc3/lib/openc3/io/posix_serial_driver.rb:151]: This jumps back to the read_nonblock right?

File: openc3/lib/openc3/io/posix_serial_driver.rb [hunk 3]
```diff
@@ -132,9 +137,19 @@ def read
...
+        begin
+          read_ready, _ = IO.fast_select(@readers, nil, nil, @read_timeout)
+        rescue IOError
+          @pipe_reader.close unless @pipe_reader.closed?
+          return ""
+        end
+        if read_ready
+          if read_ready.include?(@pipe_reader)
+            @pipe_reader.close unless @pipe_reader.closed?
+            return ""
```
Comment (jmthomas) [inline openc3/lib/openc3/io/posix_serial_driver.rb:149]: Why would you close the pipe_reader if the read_ready says it's ready?

**Review by jmthomas — COMMENTED** [review]
I'm not familiar with this technique and how it works.

**Comment by ryanmelt** [comment]
This mostly the same code from sleeper.rb.
See: https://www.sitepoint.com/the-self-pipe-trick-explained/
for the technique

File: openc3/lib/openc3/io/posix_serial_driver.rb (hunk 1 above)
Comment (ryanmelt) [reply openc3/lib/openc3/io/posix_serial_driver.rb:143]: Closing the pipe reader is just to make sure we don't leave file handles open.   Return "" means that we are dead and should no longer be used going forward.  IOError should only occur if the serial port has already been closed.

File: openc3/lib/openc3/io/posix_serial_driver.rb (hunk 3 above)
Comment (ryanmelt) [reply openc3/lib/openc3/io/posix_serial_driver.rb:149]: pipe_reader is only ever ready, if pipe_writer was written.  That means we have been closed.

File: openc3/lib/openc3/io/posix_serial_driver.rb (hunk 2 above)
Comment (ryanmelt) [reply openc3/lib/openc3/io/posix_serial_driver.rb:151]: Yes

**Review by jmthomas — APPROVED** [review]

**Pull request merged**